### ISSUE TITLE
Integration of images from Assetto Corsa

### ISSUE THREAD

**Issue by lupusorina** [dasGringuen/assetto_corsa_gym#6]
Hi,

First of all, great work! 

I was wondering if you are planning to integrate other sensor outputs, such as cameras, or if you know if this task is straightforward/feasible? I am just getting started with Assetto Corsa so any advice is very appreciated on how to add image data.

Thanks!

**Comment by dasGringuen** [comment]
Hi,

Yes, images are actually in the plan. I have a small prototype, but I haven't had time to fully integrate and test it yet. If you'd like to volunteer, it would be greatly appreciated, and I will help you. Otherwise, you may need to wait a little longer.

The plan is basically to capture the image using OpenCV in the plugin, downsample it (and optionally save the raw image), and send it through the existing UDP interface. The only issue is that we would need to include OpenCV in the Python distribution of the plugin, which is not the same one being used for training the models and receiving the data. If screen capture works in Python 3.3, then the rest should be doable.

An easier option would be to capture the image directly from the client side and not through the plugin. The problem with this approach is that the images will have a slight delay, latency and potentially out of sync with the telemetry (some testing would be requiered to check how worse it gets). Additionally, it won't work in a distributed setup where the client is running on another computer / computers.

Also, if you decide to work on this in your repo, please let me know so that I can add it to this one.

Best,
Adrian

**Comment by lupusorina** [comment]
Hi Adrian,

Thanks for letting me know. If you want to add here your preliminary work, I can definitely take over and try to help the best I can, including testing and further development.

I agree with you, capturing the images directly is not great, as the image data will be out of sync with telemetry.

Best wishes,
Sorina

**Comment by dasGringuen** [comment]
Hi Sorina,

Here are my thoughts and code. Let me know what you think or if you face any issues! I plan to put some effort into this in the near future as well.

1. Image Grabber: I tested this up to 100Hz (our baselines were running at 25Hz).

  
  ```
  import time
  import cv2
  import pygetwindow as gw
  import mss
  import numpy
  import time
  with mss.mss() as sct:
      title = "Assetto Corsa"
  
      # Find the window by title
      window = gw.getWindowsWithTitle(title)[0]  # Get the first window with the title
      window.activate()
  
      # Part of the screen to capture
      monitor = {"top": window.top, "left": window.left,
                 "width": window.width, "height": window.height}
      while "Screen capturing":
          last_time = time.time()
          # Get raw pixels from the screen, save it to a Numpy array
          img = numpy.array(sct.grab(monitor))
  
          # Display the picture
          # cv2.imshow("OpenCV/Numpy normal", img)
  
          # Display the picture in grayscale
          cv2.imshow('OpenCV/Numpy grayscale',
                      cv2.cvtColor(img, cv2.COLOR_BGRA2GRAY))
  
          print("fps: {}".format(1 / ((time.time() - last_time)+ 1e-8) ))
  
          # Press "q" to quit
          if cv2.waitKey(25) & 0xFF == ord("q"):
              cv2.destroyAllWindows()
              break
  ```

2. Python Distribution: As far as I understand, if you just install the Python distribution, then AC will use that.
[Python 3.3.5](https://legacy.python.org/download/releases/3.3.5/). Then we just need to pip install from that distribution, and the packages will be visible in the AC plugin. 

3. To send the captured images, you can simply add a new field in the update function. [Here](https://github.com/dasGringuen/assetto_corsa_gym/blob/1cbd502427c576c1c8aa3cfbbe7ef2b930a9f42b/assetto_corsa_gym/assetto-corsa-autonomous-racing-plugin/plugins/sensors_par/structures.py#L28)

4. Configuration flags should be placed [here](https://github.com/dasGringuen/assetto_corsa_gym/blob/main/assetto_corsa_gym/assetto-corsa-autonomous-racing-plugin/plugins/sensors_par/config.py) (by default, image capturing should be turned off for performance reasons).

5. The Python debug interface is very convenient to develop the plug-in part. [See](https://github.com/dasGringuen/assetto_corsa_gym/issues/2)



Adrian

**Comment by lupusorina** [comment]
Awesome. Thanks a lot! I'll take a look!

**Comment by qtweng** [comment]
Are there any updates on this? I'd like to use this platform for training a vision model.

**Comment by dasGringuen** [comment]
We haven’t had the chance to fully integrate and test it yet. If you’d like to volunteer, it would be greatly appreciated, and I’m happy to assist you. Otherwise, you might need to wait a bit longer.

**Comment by lupusorina** [comment]
@qtweng yes, I also didn't get a chance to look into this due to other priorities, but it's on my todo list.

**Comment by actharsis** [comment]
Hello, I'm trying to implement image capturing from the plugin but I met with the issue that mentioned above. Python in AC doesn't have opencv lib and I need somehow to include it. I tried to print `sys.executable` from plugin when I have python 3.3.5 installed on side and it's printed path to AC executable, so I guess AC always using internal interpreter(inside of assettocorsa\system\x64\python33.dll maybe). Also installing opencv through another interpreter also big pain because pip dropped support for python 3.3. Do you have any ideas to get rid of this problem?

**Comment by dasGringuen** [comment]
The script below works. But it takes ~10ms to grab each frame. Depending on the application, this could be enough. Maybe setting the sampling for the RL agent at 10Hz. DirectX is not an option (should be as low as 5ms) since it doesn’t work in Python 3.3, and there's no way to upgrade the distribution for Assetto Corsa.  I will integrate this in the repo later this week.

### Some options to improve this:  
We could spawn a separate process, **frame_grabber**, from `sensor_par` to handle fast screen capture. This process would continuously update a shared memory space with the latest frame. We could make the update only when `info.graphics.packetId` changes, so it stays in sync with Assetto Corsa.

The client can then grab frames directly from shared memory or, if running remotely, receive them via UDP. `frame_grabber` would handle the streaming, keeping things low-latency, whether running locally or in a distributed setup. Eventually this could be a small c++ code.

### To run the script:
We need to add some packages to the distribution in AC.

Download Python 3.3.5:  
🔗 [Python 3.3.5 Installer](http://www.python.org/ftp/python/3.3.5/python-3.3.5.amd64.msi)  
- Install it in `C:\Python33\`  
- Make sure to check **"Add Python to PATH"**  

Set up `pip`:  
```sh
curl https://bootstrap.pypa.io/pip/3.3/get-pip.py -o get-pip.py
C:\Python33\python get-pip.py
```
Install dependencies:  
```sh
C:\Python33\python -m pip --version
C:\Python33\Scripts\pip install mss==3.3.0 Pillow==3.4.2 pygetwindow
```
Check if it works:  
```sh
C:\Python33\python -c "import mss, pygetwindow; print('Success')"
```


The script, `screen_capture.py`,  to be added in the same path as `sensors_par`:

``` Python
import os
import time
import pygetwindow as gw
import ctypes
import ctypes.wintypes
from PIL import Image

import logging
logger = logging.getLogger(__name__)

# Windows API functions
user32 = ctypes.windll.user32
gdi32 = ctypes.windll.gdi32

def capture(save_to_disk=False, save_path="screenshots", scale=1.0, verbose=True):
    """
    Returns the processed image object (PIL Image).
    """

    title = "Assetto Corsa"

    # Find the Assetto Corsa window
    windows = gw.getWindowsWithTitle(title)
    if not windows:
        if verbose:
            logger.info("Assetto Corsa window not found!")
        return None

    window = windows[0]  # Get window handle
    hwnd = window._hWnd

    # Get window size
    rect = ctypes.wintypes.RECT()
    user32.GetWindowRect(hwnd, ctypes.byref(rect))
    width, height = rect.right - rect.left, rect.bottom - rect.top

    # Ensure new_width and new_height are always defined
    new_width, new_height = width, height

    # Create device contexts
    hdc = user32.GetDC(hwnd)
    mem_dc = gdi32.CreateCompatibleDC(hdc)
    bmp = gdi32.CreateCompatibleBitmap(hdc, width, height)
    gdi32.SelectObject(mem_dc, bmp)

    # Capture the window
    start_time = time.time()
    result = user32.PrintWindow(hwnd, mem_dc, 0)  # Faster than mss.grab()
    capture_time = time.time() - start_time

    if not result:
        if verbose:
            logger.info("PrintWindow() failed!")
        return None

    if verbose:
        logger.info("Screen captured in {:.4f} seconds".format(capture_time))

    if verbose:
        logger.info("Screen captured in {:.4f} seconds".format(capture_time))

    # Define bitmap info
    class BITMAPINFOHEADER(ctypes.Structure):
        _fields_ = [
            ("biSize", ctypes.c_uint32),
            ("biWidth", ctypes.c_long),
            ("biHeight", ctypes.c_long),
            ("biPlanes", ctypes.c_uint16),
            ("biBitCount", ctypes.c_uint16),  # **Set this to 16-bit RGB565**
            ("biCompression", ctypes.c_uint32),
            ("biSizeImage", ctypes.c_uint32),
            ("biXPelsPerMeter", ctypes.c_long),
            ("biYPelsPerMeter", ctypes.c_long),
            ("biClrUsed", ctypes.c_uint32),
            ("biClrImportant", ctypes.c_uint32),
        ]

    bmi = BITMAPINFOHEADER()
    bmi.biSize = ctypes.sizeof(BITMAPINFOHEADER)
    bmi.biWidth = width
    bmi.biHeight = -height  # Negative means top-down
    bmi.biPlanes = 1
    bmi.biBitCount = 16  # **Directly request RGB565 (2 bytes per pixel)**
    bmi.biCompression = 0  # BI_RGB (uncompressed)

    # Allocate buffer for pixel data
    pixel_data = ctypes.create_string_buffer(width * height * 2)  # **RGB565 is 2 bytes per pixel**

    # Extract pixels
    gdi32.GetDIBits(mem_dc, bmp, 0, height, pixel_data, ctypes.byref(bmi), 0)

    # Convert to a PIL image without extra processing
    img = Image.frombuffer("RGB", (width, height), pixel_data.raw, "raw", "BGR;16", 0, 1)

    # Apply downsampling if scale != 1.0
    if scale != 1.0:
        new_width = int(width * scale)
        new_height = int(height * scale)
        img = img.resize((new_width, new_height), Image.BILINEAR)  # Faster than ANTIALIAS
        if verbose:
            logger.info("Downsampled Image: {}x{}".format(new_width, new_height))

    # Save to disk if enabled
    if save_to_disk:
        if not os.path.exists(save_path):
            os.makedirs(save_path)

        # Generate timestamped filename
        timestamp = time.strftime("%Y%m%d_%H%M%S")
        filename = os.path.join(save_path, "ac_screenshot_{}.png".format(timestamp))

        # Save image
        img.save(filename)
        if verbose:
            logger.info("Screenshot saved: {}".format(filename))

    total_time = time.time() - start_time  # Total execution time
    if verbose:
        logger.info("Total execution time: {:.4f} seconds".format(total_time))

    # Cleanup
    user32.ReleaseDC(hwnd, hdc)
    gdi32.DeleteObject(bmp)
    gdi32.DeleteDC(mem_dc)

    return img, width, height, new_width, new_height

```

And call it from `ego_server.py` adding this line in the `tick` method:

```
if (self.total_steps % (self.config.sampling_freq // self.config.ego_sampling_freq)) == 0:
    img = screen_capture.capture(save_to_disk=False, save_path="captures", scale=1.0, verbose=False)
```

**Comment by dasGringuen** [comment]
15ms at resolution @320 x300 -> smallest that works

You can set this values in:
<>\Documents\Assetto Corsa\cfg\video.ini
[VIDEO]
HEIGHT=300
WIDTH=320

**Comment by actharsis** [comment]
I tried this method and it works, thanks. But there are significant performance issue, especially when you saving images to disk. It locks a possibility to grab the data from real driver. I wondering if it possible to grab the data not in realtime but from the replay. But I guess your idea with multiprocessing is best option in this case.

**Comment by dasGringuen** [comment]
Right, Assetto Corsa supports replays. You could use that until we have the improved version. I will give it a try soon.

**Comment by dasGringuen** [comment]
Hi!  

I pushed a new commit with screen capture support. It spawns a secondary Python process using **Python 3.9**, which captures images, writes them to a dual-buffer shared memory*, and allows the AC environment (`ac_env.py`) to access them. Additionally, images can be saved to disk.  
This feature works for both human and AI drivers. I also added functionality to allow the use of a different interpreter within the plugin (needed for running OpenCV).  

### **Performance:**  
- The implementation is highly efficient, with no noticeable FPS impact.  
- Tested at **640x480 @ 25Hz** on a slow computer while saving to disk.  
- Higher resolutions **affect performance** when grabbing the screen, but the impact when using only shared memory remains unclear and requires further testing.  
- I left some performance measurement functionalities in the plugin.  
- 
### **Next Steps:**  
- Verify installation and dependencies.  
- Test performance at different resolutions.  
- Add a benchmark policy that runs from images.  

### **Demo:**  
See **test_gym_images.ipynb** and the **main README** for a working example.  
Detailed setup instructions are available in README.md and [`INSTALL.md`](INSTALL.md).  


Let me know if you test it and/or find something that isn’t working.

Have fun,
Adrian

@lupusorina @actharsis @qtweng
